tsl/src/remote/txn.c

```diff
@@ -444,112 +506,161 @@ remote_txn_write_persistent_record(RemoteTxn *entry)
 	entry->remote_txn_id = remote_txn_persistent_record_write(entry->id);
 }
 
+static void
+on_prepare_transaction_response(AsyncRequest *req, AsyncResponse *rsp, void *data)
+{
+	bool success = on_remote_txn_response(req, rsp);
+
+	if (!success)
+	{
+		RemoteTxn *txn = data;
+
+		/* If the prepare is not successful, reset the remote transaction ID
+		 * to indicate we need to do a rollback */
+		txn->remote_txn_id = NULL;
+	}
+}
+
 AsyncRequest *
 remote_txn_async_send_prepare_transaction(RemoteTxn *entry)
 {
+	AsyncRequest *req;
+
 	Assert(entry->conn != NULL);
-	Assert(entry->xact_depth > 0);
+	Assert(remote_connection_xact_depth_get(entry->conn) > 0);
 	Assert(entry->remote_txn_id != NULL);
 
 	elog(DEBUG3,
 		 "2pc: preparing remote transaction on connection %p: %s",
 		 entry->conn,
 		 remote_txn_id_out(entry->remote_txn_id));
-	return async_request_send(entry->conn,
-							  remote_txn_id_prepare_transaction_sql(entry->remote_txn_id));
+
+	remote_connection_xact_transition_begin(entry->conn);
+	req = async_request_send(entry->conn,
+							 remote_txn_id_prepare_transaction_sql(entry->remote_txn_id));
+	async_request_set_response_callback(req, on_prepare_transaction_response, entry);
+
+	return req;
 }
 
 AsyncRequest *
 remote_txn_async_send_commit_prepared(RemoteTxn *entry)
 {
+	AsyncRequest *req;
+
 	Assert(entry->conn != NULL);
 	Assert(entry->remote_txn_id != NULL);
 
 	elog(DEBUG3,
 		 "2pc: commiting remote transaction on connection %p: '%s'",
 		 entry->conn,
 		 remote_txn_id_out(entry->remote_txn_id));
-	return async_request_send_with_error(entry->conn,
-										 remote_txn_id_commit_prepared_sql(entry->remote_txn_id),
-										 WARNING);
+
+	remote_connection_xact_transition_begin(entry->conn);
+
+	req = async_request_send_with_error(entry->conn,
+										remote_txn_id_commit_prepared_sql(entry->remote_txn_id),
+										WARNING);
+	async_request_set_response_callback(req, on_commit_or_commit_prepared_response, entry);
+
+	return req;
 }
 
+/*
+ * Rollback a subtransaction to a given savepoint.
+ */
 bool
 remote_txn_sub_txn_abort(RemoteTxn *entry, int curlevel)
 {
-	StringInfoData sql;
 	PGconn *pg_conn = remote_connection_get_pg_conn(entry->conn);
+	bool success = true;
 
-	Assert(entry->xact_depth == curlevel);
-	Assert(entry->xact_depth > 1);
-	initStringInfo(&sql);
-
-	if (in_error_recursion_trouble())
-		return false;
+	Assert(remote_connection_xact_depth_get(entry->conn) == curlevel);
+	Assert(remote_connection_xact_depth_get(entry->conn) > 1);
 
-	if (PQtransactionStatus(pg_conn) != PQTRANS_INTRANS &&
-		PQtransactionStatus(pg_conn) != PQTRANS_INERROR)
-		return false;
+	if (in_error_recursion_trouble() && remote_connection_xact_is_transitioning(entry->conn))
+		remote_connection_xact_transition_begin(entry->conn);
```

Comment:
Something is wrong here. The condition is satisfied if transitioning already started. Should it be:
```suggestion
	if (in_error_recursion_trouble() && !remote_connection_xact_is_transitioning(entry->conn))
		remote_connection_xact_transition_begin(entry->conn);
```